igvm/hypervisor_preferences.py

```diff
@@ -66,209 +96,336 @@ def __repr__(self):
         cur_hv_cpus = sum(
             v[self.attribute] for v in vm.hypervisor.dataset_obj['vms']
         )
         cur_hv_rl_cpus = vm.hypervisor.dataset_obj[self.attribute]
         cur_ovr_allc = float(cur_hv_cpus) / float(cur_hv_rl_cpus)
 
+        # Calculate by how much we would overbook the target HV.
         tgt_hv_cpus = vm.dataset_obj[self.attribute] + sum(
             v[self.attribute] for v in hv.dataset_obj['vms']
         )
         tgt_hv_rl_cpus = hv.dataset_obj[self.attribute]
         tgt_ovr_allc = float(tgt_hv_cpus) / float(tgt_hv_rl_cpus)
 
-        return tgt_ovr_allc > cur_ovr_allc
-
+        # Whether the target hv would be more overbooked than the current one.
+        rel_overbooking = tgt_ovr_allc / cur_ovr_allc
+        if rel_overbooking > 1.:
+            return .01
 
-class HashDifference(object):
-    """Return some arbitrary number to have stable ordering"""
-    def __repr__(self):
-        return '{}()'.format(type(self).__name__)
+        # Normalize the value. This is no hard criteria, except for non-
+        # overbookable resources like disk. Hence we can allow overbooked cpus.
+        # For cpus the avg load is more important, for anything else non-
+        # overbookable we will leave the corresponding hard-check for later.
+        return 1 - rel_overbooking
 
-    def __call__(self, vm, hv):
-        return hash(hv.fqdn) - hash(vm.fqdn)
 
+class PreferenceEvaluator(object):
+    """Evaluates all preferences for a given VM and HV and calculates the total
+    score based on which the most preferred HVs can be picked.
+    """
+    def __init__(self, preferences: List[HypervisorPreference]) -> None:
+        self.preferences = preferences
+
+    def get_total_score(self, vm, hv) -> float:
+        """Calculates the total score for a given VM and HV pair."""
+        n_prefs = len(self.preferences)
+        matched_prefs = 0
+        sum_prefs = 0.
+
+        log.debug('Checking {}..'.format(str(hv)))
+
+        # Checking HV against all preferences.
+        for pref in self.preferences:
+            result = float(pref.get_score(vm, hv))
+
+            # We expect normalized values from 0 - 1.
+            if result < 0. or result > 1.:
+                raise ValueError(
+                    'Preference must be expressed in a 0.0 - 1.0 range, '
+                    '{} given.'.format(result)
+                )
+
+            # Add up the individual preference scores.
+            if result > 0.:
+                log.debug('Preference {} matches with score {:.4f}.'.format(
+                    str(pref),
+                    result,
+                ))
+
+                matched_prefs += 1
+                sum_prefs += result
+            else:
+                log.debug(
+                    'Hypervisor "{}" is skipped because preference "{}" does '
+                    'not match.'.format(str(hv), str(pref)),
+                )
+
+        # Exclude HV if only one criteria has failed.
```

Comment:
If at least one? If even one? If any?

Reply:
if any of the preferences fails, then yes. hv will be excluded

Reply:
Then please correct the way that sentence is written.

Reply:
fixed